integration/integration_test.go

```diff
@@ -66,25 +67,58 @@ func TestHttpIngress(t *testing.T) {
+					assert.Equal(t, []string{"Header from FTL"}, resp.headers["Get"])
+
+					message, ok := resp.body["msg"].(string)
 					assert.True(t, ok, "message is not a string")
 					assert.Equal(t, "UserID: 123, PostID: 456", message)
 
+					nested, ok := resp.body["nested"].(map[string]any)
+					assert.True(t, ok, "nested is not a map")
+					goodStuff, ok := nested["good_stuff"].(string)
+					assert.True(t, ok, "good_stuff is not a string")
+					assert.Equal(t, "This is good stuff", goodStuff)
 				}),
-				httpCall(rd, http.MethodPost, "/users", obj{"userID": "123", "postID": "345"}, func(t testing.TB, resp *httpResponse) {
+				httpCall(rd, http.MethodPost, "/users", jsonData(t, obj{"userID": 123, "postID": 345}), func(t testing.TB, resp *httpResponse) {
 					assert.Equal(t, 201, resp.status)
+					assert.Equal(t, []string{"Header from FTL"}, resp.headers["Post"])
+					success, ok := resp.body["success"].(bool)
+					assert.True(t, ok, "success is not a bool")
+					assert.True(t, success)
 				}),
 				// contains aliased field
-				httpCall(rd, http.MethodPost, "/users", obj{"id": "123", "postID": "345"}, func(t testing.TB, resp *httpResponse) {
+				httpCall(rd, http.MethodPost, "/users", jsonData(t, obj{"user_id": 123, "postID": 345}), func(t testing.TB, resp *httpResponse) {
 					assert.Equal(t, 201, resp.status)
 				}),
-				httpCall(rd, http.MethodPut, "/users/123", obj{"postID": "346"}, func(t testing.TB, resp *httpResponse) {
+				httpCall(rd, http.MethodPut, "/users/123", jsonData(t, obj{"postID": "346"}), func(t testing.TB, resp *httpResponse) {
+					assert.Equal(t, 200, resp.status)
+					assert.Equal(t, []string{"Header from FTL"}, resp.headers["Put"])
+					assert.Equal(t, map[string]any{}, resp.body)
+				}),
+				httpCall(rd, http.MethodDelete, "/users/123", jsonData(t, obj{}), func(t testing.TB, resp *httpResponse) {
 					assert.Equal(t, 200, resp.status)
+					assert.Equal(t, []string{"Header from FTL"}, resp.headers["Delete"])
+					assert.Equal(t, map[string]any{}, resp.body)
 				}),
-				httpCall(rd, http.MethodDelete, "/users/123", obj{}, func(t testing.TB, resp *httpResponse) {
+
+				httpCall(rd, http.MethodGet, "/html", jsonData(t, obj{}), func(t testing.TB, resp *httpResponse) {
+					assert.Equal(t, 200, resp.status)
+					assert.Equal(t, []string{"text/html; charset=utf-8"}, resp.headers["Content-Type"])
+					assert.Equal(t, "<html><body><h1>HTML Page From FTL 🚀!</h1></body></html>", string(resp.bodyBytes))
+				}),
+
+				httpCall(rd, http.MethodPost, "/bytes", []byte("Hello, World!"), func(t testing.TB, resp *httpResponse) {
 					assert.Equal(t, 200, resp.status)
+					assert.Equal(t, []string{"application/octet-stream"}, resp.headers["Content-Type"])
+
+					respBodyStr := strings.Trim(string(resp.bodyBytes), "\"")
+					decodedBody, err := base64.StdEncoding.DecodeString(respBodyStr)
+					if err != nil {
+						t.Fatalf("Failed to decode base64: %v", err)
+					}
```

Comment:
@alecthomas maybe I should change the behavior of the `[]byte` response to already be decoded? So that if we send bytes to an ingress and return them, they should match? Whatcha think?